[go _route] fragment parameter added

AffanShaikhsurab opened the pull request:
### Description of Change

This PR addresses the need for fragments/hashes to be treated as first-party parameters within the `go_router` package, as highlighted in [issue #150155](https://github.com/flutter/flutter/issues/150155). Previously, users had to manually append the fragment to the URL, which could lead to potential bugs. With this update, the `fragment` is now a dedicated parameter, allowing for a more seamless and bug-free integration.

#### Before:

```dart
final location = context.namedLocation('some_route'); // const nested records
// Manually adding the fragment, hoping there aren't any weird bugs surrounding it
context.replace('$location#https://a.url/that?i=mightuse');
```

#### After:

```dart
// Directly passing the fragment as a parameter
context.goNamed("details", fragment: 'https://a.url/that?i=mightuse');

// or

final location = GoRouterState.of(context).namedLocation(
  'details',
  fragment: 'section3',
);
context.go(location);
```

### Issues Fixed

This PR resolves [issue #150155](https://github.com/flutter/flutter/issues/150155).

---

## Pre-launch Checklist

- [✔️ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [✔️ ] I read the [Tree Hygiene] page, which explains my responsibilities.
- [✔️ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [✔️ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [✔️ ] I [linked to at least one issue that this PR fixes] in the description above.
- [✔️ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ✔️] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [✔️ ] I updated/added relevant documentation (doc comments with `///`).
- [ ✔️] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ✔️] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

AffanShaikhsurab commented:
Can someone please tell how to solve this Linux repo_checks !

stuartmorgan commented:
> Can someone please tell how to solve this Linux repo_checks !

If you click through the Details link, you can see that it's failing the format check. You've checked this box:
>> * [✔️ ] I read and followed the [relevant style guides](https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style) and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)

but the code in your PR does not match `dart format` output.

AffanShaikhsurab commented:
> > Can someone please tell how to solve this Linux repo_checks !
> 
> If you click through the Details link, you can see that it's failing the format check. You've checked this box:
> 
> > > * [✔️ ] I read and followed the [relevant style guides](https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style) and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
> 
> but the code in your PR does not match `dart format` output.

Sorry as I am new I don't know much to be honest . I ran the dart format but nothing changed can you please point me to some resource so I can learn more about formatting :: 
![image](https://github.com/user-attachments/assets/65e3e1ce-640c-46ec-bb0a-28760a86c592)

AffanShaikhsurab commented:
I am getting this error  from another package , what should I do:: 

```
ERROR: Pub does not support publishing packages with non-resolving symlink: `/b/s/w/ir/x/w/packages/packages/in_app_purchase/in_app_purchase_storekit/example/macos/RunnerTests/Stubs.h` => `../../ shared / RunnerTests / Stubs.h`.

Unable to publish in_app_purchase_storekit

[packages/in_app_purchase/in_app_purchase_storekit completed in 0m 4s
```

stuartmorgan commented:
> I am getting this error from another package , what should I do

Your PR [destroys a number of symlinks](https://github.com/flutter/packages/pull/7434/files#diff-5b1bbb228d85dd9881bb37029dc4a597978b6dcb03da4452939f82fe6a066eceR1). You PR needs to not damage files in other packages.

Marking as a draft pending a reviewable PR.

stuartmorgan commented:
Note that you've also checked this box:

> * [ ✔️] I added new tests to check the change I am making, or this PR is [test-exempt](https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests).

but the PR does not contain any new tests, which will not pass review.

> Sorry as I am new I don't know much to be honest .

I strongly recommend carefully reading and following the checklist in the PR; helping new contributors understand what all the required steps are is one of its main purposes.

AffanShaikhsurab commented:
yeah sure, I will try my best to follow the checklist. Thanks!

AffanShaikhsurab commented:
I have[ added the parameters ](https://github.com/flutter/packages/pull/7434/files#diff-68f73e9c61b4ce3281bda2f29796998d510c32f83170985404a12700fd3dfc29)in the test-helper , Am i supposed to write a separate test-case ?

stuartmorgan commented:
> Am i supposed to write a separate test-case ?

Which existing test would fail if all the code handling the new parameter were removed or broken?

onbreadassist commented:
This namedLocation and goNamed : https://github.com/flutter/packages/pull/7434/files#diff-68f73e9c61b4ce3281bda2f29796998d510c32f83170985404a12700fd3dfc29

AffanShaikhsurab commented:
This namedLocation and goNamed : https://github.com/flutter/packages/pull/7434/files#diff-68f73e9c61b4ce3281bda2f29796998d510c32f83170985404a12700fd3dfc29

stuartmorgan commented:
Those aren't tests, they are utilities to write tests with.

What existing tests would fail if you removed [this code](https://github.com/flutter/packages/pull/7434/files#diff-b1ee8945e4afd90def0a63ab64668aac55b69f11719b3e1c25aa6a85f4a4d6f7R370-R373) and/or reverted [this section](https://github.com/flutter/packages/pull/7434/files#diff-999d3b39cc560038cc4fa71db72f4b8bc142217ce0dbd5d009dfc4a9a64e8425R21-R26)?

stuartmorgan commented:
> * [✔️ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy](https://dart.dev/tools/pub/versioning),

You are adding new public API, which is a `minor` change according to the linked versioning philosophy, but you haven't changed the `minor` version component.

> * [ ✔️] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style](https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style), or this PR is [exempt from CHANGELOG changes](https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog).

Your changes need to follow the linked style guide.

AffanShaikhsurab commented:
if i remove ` GoRouter.of(this).namedLocation(
          fragment == null ? name : '$name#$fragment',
          pathParameters: pathParameters,
          queryParameters: queryParameters);
` this is the output:`00:28 +0 -27: Some tests failed.`

AffanShaikhsurab commented:
> > * [✔️ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy](https://dart.dev/tools/pub/versioning),
> 
> You are adding new public API, which is a `minor` change according to the linked versioning philosophy, but you haven't changed the `minor` version component.
> 
> > * [ ✔️] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style](https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style), or this PR is [exempt from CHANGELOG changes](https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog).
> 
> Your changes need to follow the linked style guide.

Understood

chunhtai reviewed (commented):
It seems like this pr have added some unrelated file or changes, can you revert those? To write a test, you can add an `testWidgets` entry in packages/packages/go_router/test/go_router_test.dart by, you should be able to follow some example in that file to set up the test. Once you done, you can run `flutter test <test_file_path>` in the go_router root folder to verify it passes

AffanShaikhsurab commented:
> It seems like this pr have added some unrelated file or changes, can you revert those? To write a test, you can add an `testWidgets` entry in packages/packages/go_router/test/go_router_test.dart by, you should be able to follow some example in that file to set up the test. Once you done, you can run `flutter test <test_file_path>` in the go_router root folder to verify it passes

I have tried creating a test case! Please check if its correct ?

chunhtai commented on .gitignore:
```diff
@@ -5,6 +5,7 @@
 .pub/
 .dart_tool/
 .vscode/
+.history/
```
can you revert this?

chunhtai commented on packages/go_router/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+## 14.3.0
+
+  - Adds support for handling fragments/hashes as a first-party parameter in the GoRouter package.
```
I think you can have one entry instead of 3

something like

Adds the fragment support to relevant GoRouter APIs

chunhtai commented on packages/go_router/lib/src/route.dart:
````diff
@@ -428,8 +429,25 @@ class GoRoute extends RouteBase {
   /// ```
+  final String? fragment;
+
+  /// The optional fragment/hash component of the route.
````
This doc string seems to be in the wrong place

chunhtai commented on packages/go_router/lib/src/route.dart:
```diff
@@ -269,6 +269,7 @@ class GoRoute extends RouteBase {
     super.redirect,
     this.onExit,
     super.routes = const <RouteBase>[],
+    this.fragment,
```
why is this needed?

AffanShaikhsurab replied:
You are right , I have removed the fragment part from the file!. Sorry for the inconvenience and Thanks for supporting

chunhtai reviewed (commented):
Can you revert changes that are not related to this pr?

chunhtai commented:
I can't review the pr as is since it includes a lot of unrelated change that should be reverted first. Can you revert the change and ping me for review again?

AffanShaikhsurab commented:
Yeah sure

chunhtai commented:
@AffanShaikhsurab there are still a bunch of unrelated file changes...
